### ISSUE TITLE
Neuralthread - cannot compile for Delphi 10.4.2 for LINUX (popos/ubuntu 20.04)

### ISSUE THREAD

**Issue by laurenceliew** [joaopauloschuler/neural-api#56]
Hi

The code runs perfectly under Win64. But I get an error on for Linux.

I am trying to compile the same SimpleMNIST demo for Linux (20.04 PopOS/Ubuntu). 

[DCC Fatal Error] neuralthread.pas(66): F2613 Unit 'Windows' not found.

uses
  Classes, SysUtils,
  {$IFDEF FPC}
  fgl, MTPCPU
    {$IFDEF WINDOWS}
    ,windows
    {$ELSE}
    ,BaseUnix
    {$ENDIF}
  {$ELSE}
  Generics.Collections, Windows  <---- E2613
  {$ENDIF}
  , syncobjs;

**Comment by joaopauloschuler** [comment]
I regret to say that I don't have Delphi for Linux to test (I test with Lazarus). Do you know what is the equivalent unit with Delphi for Linux?
If you propose a fix, I'll be glad to add.

**Comment by laurenceliew** [comment]
We can just comment it out, but other errors for the TRLCriticalSections which I think is not available in Delphi Linux pops up. I think we need to find the equivalent for Linux.  I am really rusty with Delphi and just getting back... so can't be of help now with these errors.

![image](https://user-images.githubusercontent.com/22905354/116806356-02abfa00-ab5f-11eb-87b3-02129a8189ac.png)

**Comment by joaopauloschuler** [comment]
It seems that Delphi for Linux has other functions for mutex:
https://www.embeddedcomputing.com/technology/software-and-os/thread-synchronization-in-linux-and-windows-systems-part-3

Would you like to give a go with Lazarus 64 for Linux?
https://www.lazarus-ide.org/

Also, if you like, you can give a go via web browser:
https://colab.research.google.com/github/joaopauloschuler/neural-api/blob/master/examples/SimpleImageClassifier/SimpleImageClassifierCPU.ipynb

In the case that you decide to try any of the above, I'm curious to know how you go.

**Comment by laurenceliew** [comment]
OK - I have managed to get your code to compile and run - but it exit with 

lliew@pop-os:~/PAServer/scratch-dir/lliew-Pop20/SimpleMNist$ ./SimpleMNist
Creating Neural Network...
File:train Labels:60000 Images:60000 Rows:28 Cols:28
File:t10k Labels:10000 Images:10000 Rows:28 Cols:28
Exception ESyncObjectException in module SimpleMNist at 000000000047D781.
Named synchronization objects not supported on this platform.

Further investigation shows that your code is using "Named synchronization" - which Linux do not support. I assume your conditional compilation fix this issue with Lazarus on Linux.

The fix for neuralthread.pas is shown below. It will allow you code to compile (but not run successfully due to the above). The fix came from Log4D code.


uses
  Classes,Generics.Collections,
//  {$IFDEF FPC}
//  fgl, MTPCPU
//    {$IFDEF WINDOWS}
//    ,windows
//    {$ELSE}
//    ,BaseUnix
//    {$ENDIF}
//  {$ELSE}
//  Generics.Collections
//  {$ENDIF}
//  , syncobjs;

 {$IFDEF LINUX64}
 Syncobjs,
 {$ELSE}
 Syncobjs, Windows,
 {$ENDIF}
 SysUtils;


type

{$IFDEF LINUX64}
  TRTLCriticalSection = TCriticalSection;
{$ENDIF}

{$IFDEF LINUX64}
  procedure EnterCriticalSection(var CS: TCriticalSection);
  procedure LeaveCriticalSection(var CS: TCriticalSection);
  procedure InitializeCriticalSection(var CS: TCriticalSection);
  procedure DeleteCriticalSection(var CS: TCriticalSection);
  function GetCurrentThreadID: Integer;
{$ENDIF}



implementation

{$IFDEF LINUX64}
procedure EnterCriticalSection(var CS: TCriticalSection);
begin
  CS.Enter;
end;

procedure LeaveCriticalSection(var CS: TCriticalSection);
begin
  CS.Leave;
end;

procedure InitializeCriticalSection(var CS: TCriticalSection);
begin
  CS := TCriticalSection.Create;
end;

procedure DeleteCriticalSection(var CS: TCriticalSection);
begin
  CS.Free;
end;

function GetCurrentThreadID: Integer;
begin
  Result := 0;
end;

{$ENDIF}

**Comment by laurenceliew** [comment]
Ok - I found the code to make the thread non-named. The code (SimpleMNIST) is running now and seems to be working.  I will tests other Demos and update later.

constructor TNeuralThread.Create(CreateSuspended: boolean; pIndex: integer);
var
  NStartName, NFinishName: string;
  PidAndIndexStr: string;
begin
  inherited Create(CreateSuspended);
  FProc := nil;
  FIndex := pIndex;
  FThreadNum := 1;
  FShouldStart := false;
  FProcFinished := false;
  PidAndIndexStr := IntToStr(GetProcessId())+'-'+IntToStr(pIndex);
  NStartName := 'NStart-'+PidAndIndexStr;
  NFinishName := 'NFinish-'+PidAndIndexStr;
  {$IFDEF FPC}
  FNeuronStart := TEventObject.Create(nil, True, False, NStartName) ;
  FNeuronFinish := TEventObject.Create(nil, True, False, NFinishName) ;
  {$ELSE}
    {$IFDEF LINUX64}
      FNeuronStart := TEvent.Create(nil, True, False, '') ;
      FNeuronFinish := TEvent.Create(nil, True, False,'') ;
    {$ELSE}
      FNeuronStart := TEvent.Create(nil, True, False, NStartName) ;
      FNeuronFinish := TEvent.Create(nil, True, False, NFinishName) ;
     {$ENDIF}
  {$ENDIF}
end;

**Comment by joaopauloschuler** [comment]
Well done @laurenceliew ! You can either create a branch and then send a pull request or send the updated file as an attachment. If you prefer the branch/pull request, your name will be stored in the commit history.

**Comment by laurenceliew** [comment]
ok.. SimpleMNISt completed and seems to run fine. This is running in a VM with 2 cores and 4GB Ram. I will now try it on my PC (need to reinstall with Ubuntu)... and the rest of the Demos. This PC has a Nvidia GPU and will test the GPU acceleration also.  If all is working... will then send over the changes.

![image](https://user-images.githubusercontent.com/22905354/116967273-8511f680-ace4-11eb-915a-a086318daf6f.png)

**Comment by joaopauloschuler** [comment]
Cool. You'll need opencl drivers. It should be similar to this (changing version number):
install nvidia-384 nvidia-opencl-icd-384 clinfo

Then, you'll need the ocl:
apt-get install ocl-icd-opencl-dev

**Comment by laurenceliew** [comment]
Thank You. The info is helpful.

--
Laurence @ PostboxSG

**Comment by joaopauloschuler** [comment]
Most probably, you'll be able to find the current installed version with dpkg -L | greep nvidia

Then, you'll be able to install the compatible driver nvidia-opencl-icd-XXX .

To be sure that the installation worked, you may run "clinfo" and check if any available device shows up:
http://manpages.ubuntu.com/manpages/xenial/man1/clinfo.1.html

**Comment by laurenceliew** [comment]
I can confirm SimpleMNIST works on Linux when compiled with Delphi Linux.  This is running non openCL on a Ryzen 4700 (8c/16T).

![image](https://user-images.githubusercontent.com/22905354/117151965-259f0e00-adec-11eb-83f2-b92bb5ed3faf.png)

![image](https://user-images.githubusercontent.com/22905354/117152210-64cd5f00-adec-11eb-9986-17e2387f6c73.png)

**Comment by joaopauloschuler** [comment]
Love looking at your screenshots.
This is my output from Lazarus without AVX and without OpenCL:
```10240 Examples seen. Accuracy: 0.3632 Error: 0.58697 Loss: 0.57086 Threads: 4 Forward time: 0.48s Backward time: 0.28s Step time: 0.63s
11520 Examples seen. Accuracy: 0.4015 Error: 0.57254 Loss: 0.57471 Threads: 4 Forward time: 0.38s Backward time: 0.38s Step time: 0.62s
12800 Examples seen. Accuracy: 0.4400 Error: 0.59346 Loss: 0.58933 Threads: 4 Forward time: 0.08s Backward time: 0.68s Step time: 0.62s
14080 Examples seen. Accuracy: 0.4786 Error: 0.37746 Loss: 0.39654 Threads: 4 Forward time: 0.32s Backward time: 0.44s Step time: 0.61s
```
With AVX2, this is what I get:
```32000 Examples seen. Accuracy: 0.8059 Error: 0.10319 Loss: 0.09521 Threads: 4 Forward time: 0.41s Backward time: 0.33s Step time: 0.60s
33280 Examples seen. Accuracy: 0.8179 Error: 0.22259 Loss: 0.31650 Threads: 4 Forward time: 0.39s Backward time: 0.17s Step time: 0.62s
34560 Examples seen. Accuracy: 0.8287 Error: 0.30109 Loss: 0.30324 Threads: 4 Forward time: 0.43s Backward time: 0.14s Step time: 0.61s
```
Your results make me wonder, are you running with debug info? This is my epoch time:
```
Epochs: 1 Examples seen:60000 Validation Accuracy: 0.9704 Validation Error: 0.0889 Validation Loss: 0.0951 Total time: 0.58min
Epoch time: 0.5500 minutes. 20 epochs: 0.1833 hours.
```

**Comment by joaopauloschuler** [comment]
On your hardware (with CPU only), I would expect from 6 to 8 epochs per minute. With GPU, it will be even faster. There is a bottleneck somewhere.

**Comment by joaopauloschuler** [comment]
Just thought about something: you can try to increase the batch size to 512 (you'll get some warning messages but you can disconsider):
```
NeuralFit.Fit(NN, ImgTrainingVolumes, ImgValidationVolumes, ImgTestVolumes, {NumClasses=}10, {batchsize=}512, {epochs=}20);
```
From memory, I saw some linux machines struggling with threading. Increasing the batch size will make the threading coordination less frequent.

**Comment by laurenceliew** [comment]
Hi @joaopauloschuler  - haha yes I was in DEBUG mode. Switching to RELEASE.
![Screenshot from 2021-05-06 10-57-26](https://user-images.githubusercontent.com/22905354/117235870-21a9d500-ae5a-11eb-80eb-e343717b05e2.png)

**Comment by joaopauloschuler** [comment]
It's better! I still think that you should try to change the batch size to 512.

**Comment by laurenceliew** [comment]
batch size of 512 is much slower. Anyways not able to get OpenCL to work. Downloaded your PasOpenCL, made some minor IFDEF.. compile and can run - but access violations.

**Comment by laurenceliew** [comment]
Hi @joaopauloschuler 

attached are the 3 files which need some changes to make them work with Delphi Linux (and Windows)
and also the SimpleMNIST made to work as a console program in Delphi Linux (and Windows)

[TestNN.zip](https://github.com/joaopauloschuler/neural-api/files/6432377/TestNN.zip)
[neural.zip](https://github.com/joaopauloschuler/neural-api/files/6432383/neural.zip)

Cheers!

**Comment by joaopauloschuler** [comment]
With a bigger batch size, the step time will be bigger. In the plus side, the epoch time is likely to be smaller as there is less threading coordination.
About your changes, it would be fantastic if you could fork, apply your changes, commit and then push. This would make easier for other users to see your change plus will add you as a contributor once your commits are merged.

**Comment by joaopauloschuler** [comment]
I have never tried MNIST with a GPU. Will give a go at my end and let you know.

**Comment by laurenceliew** [comment]
Hi @joaopauloschuler 

Sorry for the miscommunication. For OpenCL - I was trying to run the SimpleImageClassiferGPU - not SimpleMNIST .
